### DISCUSSION TITLE
Minimum code needed to add some buttons to my SkiaSharp game

### DISCUSSION THREAD

**Discussion by ToolmakerSteve** [AvaloniaUI/Avalonia#16959] [Q&A] [answered]
I am writing a game, using SkiaSharp to render.

Doesn't need much in the way of traditional UI - most actions occur by user touching something they see in the world map.

I bring up a panel across the bottom, when there is a user decision better made with buttons and associated text.

Currently I do that via Maui. I ran into a *cough* bug *cough*, that I failed to work-around. Also an issue where player could click on a button, but it wasn't ready yet so nothing happened [switching between my Skia game loop and Maui -- I can't blame Maui for this one].

This got me thinking: do I need **any** UI framework? If it was ALL part of my Skia game loop, and the rendering was all my code, I would have gotten it working by now.

It seemed silly to reinvent the wheel; I figured *someone* must have written a **lightweight widget package for SkiaSharp**.  
I haven't found one yet.

Then I thought of Avalonia. Much, much more than I need. But contains code for widgets, and renders to Skia.

So my idea is to copy a bare minimum of files that would render buttons, and handle user interaction with those buttons.

Then decide how best to alter my "game loop" to include those buttons. Or the opposite: learn Avalonia's approach, alter my code.

-----------------------------------

**Suggestions on what parts of Avalonia code I need?**

* I see `Avalonia.Skia` - that is one obvious place to start.

* I will build and run whatever Avalonia sample seems closest to my need. Tracing through that code will eventually tell me what I need to know. But my concern is that I might have to trace through a lot of code that I won't need.

* Layout, measurement, building a visual tree: there must be a lot of generality that I could skip. I might replace layout management with a simpler one.

* I won't have any XAML. I guess c# markup is worthwhile. Cleaner than my current code.

* I have a global scale to apply.

* Stacking elements vertically or horizontally is a good building block.

* I haven't thought much about *input*. So far its simply Skia *touch*.

**Comment by thevortexcloud** [comment]
If you want to gut Avalonia, you would be better off just writing your own simple UI system. If you use something like Mono Game/SDL, it will do a lot of the hard work for you (EG creating a window and cross platform handling of input). 

There is already a C++ example for integrating regular Skia into SDL.

https://chromium.googlesource.com/skia/+/chrome/m53/example/SkiaSDLExample.cpp

Making a (square) button work is as simple as checking if the pointing device was pressed within a given range of coordinates. If you have a vague idea about how to make a UI framework, you can throw a very simple one together within a day or so.

**Reply by ToolmakerSteve** [reply]
Thanks for the suggestions.

**Reply by thevortexcloud** [reply]
You are welcome. I actually wrote my own UI system using only SkiaSharp (although it's incomplete). But I designed it for specialised hardware displays and input. So it just draws directly to the display without worrying about things like windows. It has support for layouts and custom controls. I am happy to share some sample code you can use for some ideas. But it is in need of a lot more work before it can be used as a general UI framework.

**Reply by ToolmakerSteve** [reply, marked as answer]
**tl;dr** Decided not to attempt what the original post suggested (stripping down Avalonia). Found an existing SkiaSharp-rendered-UI-widgets project to use w/ Maui, avoiding Maui's native widgets.

I agree with your comment about the folly of stripping down Avalonia (or any large framework).

Thanks for offering to share your code. Hold that thought -  
Now I found [DrawnUI.Maui](https://github.com/taublast/DrawnUi.Maui).  
And a [youtube](https://www.youtube.com/watch?v=g38f8otTT3k).

Simple layout engine; renders everything to one Skia canvas. That fits well with what I am already doing.

I'll keep Maui around. But barely use it. Get the game to a good prototype, before concerning myself with app size.  
Use almost none of Maui. One Window. One SKCanvas.
Maui.Essentials, which lives independently of Maui's UI, for other platform features.

Develop a good game prototype, before stressing about app size.

Meets my requirements. None of the "use native widgets" problems of Maui.  

Cross-platform via .Net 8, SkiaSharp, Maui.Essentials, a tiny bit of Maui. Hoping this is a good way to leverage Microsoft technologies, for an app that does not need a lot of complex UI, while avoiding their failure to offer a way to do cross-platform UI w/o dependency on each platform's UI subsystem.
